### PR TITLE
Issues/144

### LINKED ISSUE #144: Currency support
Coinmarkercap has support for many currencies at the moment, we could use that

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

### LINKED ISSUE #160: Add support for GBP currency

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
This is branch for review of the approach.

Fix #144  Fix #160

### CODE REVIEW

**Comment by OlegGordiichuk** [comment]
@michaelScoff could you pls add endpoints to the images with example and review my changes?

**Comment by michaelScoff** [comment]
@OlegGordiichuk can you use endpoint that I sent you? Here is docs for it: 
https://github.com/TrustWallet/trust-ray/wiki/API#get-prices

It's optimized to be used on mobile, since returned only tokens that you need and information that's being used in the app.

I think model for the coin could be just name, symbol, price right? 
What's the best way to handle it when you switch currency on the phone? I will add this ability in settings next week


Regarding images, I will have endpoint pretty soon! I will update it

This is a great improvement!

File: Trust/EtherClient/CoinMarket/Types/CoinTicker.swift [hunk 1]
```diff
@@ -1,12 +1,104 @@
...
+        let btcPrice: String = try container.decode(String.self, forKey: .btcPrice)
+        let volumeUsd24h: String? = try container.decodeIfPresent(String.self, forKey: .volumeUsd24h)
+        let marketCapUsd: String? = try container.decodeIfPresent(String.self, forKey: .marketCapUsd)
+        let availableSupply: String? = try container.decodeIfPresent(String.self, forKey: .availableSupply)
+        let totalSupply: String? = try container.decodeIfPresent(String.self, forKey: .totalSupply)
+        let percentChange1h: String? = try container.decodeIfPresent(String.self, forKey: .percentChange1h)
+        let percentChange24h: String? = try container.decodeIfPresent(String.self, forKey: .percentChange24h)
+        let percentChange7d: String? = try container.decodeIfPresent(String.self, forKey: .percentChange7d)
+        let lastUpdated: String = try container.decode(String.self, forKey: .lastUpdated)
+        let img: String = CoinTicker.path + id + ".png"
```
Comment (michaelScoff) [inline Trust/EtherClient/CoinMarket/Types/CoinTicker.swift]: I don't think saving image URL is a good idea, it's better to construct it on the fly

File: Trust/EtherClient/CoinMarket/CoinMarketClient.swift [hunk 2]
```diff
@@ -3,36 +3,31 @@
...
 
-enum CoinMarketService {
-    case price(id: String, currency: String)
-    case prices(limit: Int)
+enum TrustMarketService {
+    case prices(currency: Currency, symbols: [String])
 }
 
-extension CoinMarketService: TargetType {
+extension TrustMarketService: TargetType {
```
Comment (michaelScoff) [inline Trust/EtherClient/CoinMarket/CoinMarketClient.swift]: Use `TrustService` instead

File: Trust/Tokens/Types/TokensDataStore.swift [hunk 3]
```diff
@@ -152,10 +152,12 @@ class TokensDataStore {
...
 
     func updatePrices() {
-        provider.request(.prices(limit: 150)) { result in
+        var symbols = objects.map { $0.symbol }
+        symbols.append("ETH")
+        provider.request(.prices(currency: Config().currency, symbols: symbols)) { result in
             guard  case .success(let response) = result else { return }
             do {
-                let tickers = try response.map([CoinTicker].self)
+                let tickers = try response.map([CoinTicker].self, atKeyPath: "response", using: JSONDecoder())
```
Comment (michaelScoff) [inline Trust/Tokens/Types/TokensDataStore.swift]: Can you create a class similar to `ArrayResponse`?

File: Trust/EtherClient/CoinMarket/CoinMarketClient.swift (hunk 2 above)
Comment (OlegGordiichuk) [reply Trust/EtherClient/CoinMarket/CoinMarketClient.swift]: Done

File: Trust/Settings/ViewControllers/SettingsViewController.swift [hunk 4]
```diff
@@ -60,6 +60,24 @@ class SettingsViewController: FormViewController {
...
+            
+            <<< PushRow<String> {
+                $0.title = NSLocalizedString("settings.currency.button.title", value: "Currency", comment: "")
+                $0.options = viewModel.currency
+                $0.value = config.currency.rawValue
+                $0.displayValueFor = { value in
+                    return value
+                }
+            }.onChange { row in
+                self.config.currency = Currency(rawValue: row.value ?? "USD")!
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Settings/ViewControllers/SettingsViewController.swift]: Done

File: Trust/Tokens/Types/TokensDataStore.swift [hunk 5]
```diff
@@ -152,10 +152,12 @@ class TokensDataStore {
     }
 
     func updatePrices() {
-        provider.request(.prices(limit: 150)) { result in
+        var symbols = objects.map { $0.symbol }
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Tokens/Types/TokensDataStore.swift]: Done

File: Trust/Tokens/Types/TokensDataStore.swift [hunk 6]
```diff
@@ -152,10 +152,12 @@ class TokensDataStore {
...
+        symbols.append("ETH")
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Tokens/Types/TokensDataStore.swift]: Done

File: Trust/Tokens/ViewModels/TokenViewCellViewModel.swift [hunk 7]
```diff
@@ -57,14 +73,21 @@ struct TokenViewCellViewModel {
...
+        return UIFont.systemFont(ofSize: 13, weight: .regular)
     }
 
     var backgroundColor: UIColor {
         return .white
     }
 
-    var image: UIImage? {
+    var placeHolder: UIImage? {
         return R.image.ethereumToken()
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Tokens/ViewModels/TokenViewCellViewModel.swift]: Done

File: Trust/Tokens/ViewModels/TokensViewModel.swift [hunk 8]
```diff
@@ -28,7 +28,8 @@ struct TokensViewModel {
         guard let tickers = tickers else { return 0 }
         guard !token.valueBigInt.isZero, let price = tickers[token.symbol] else { return 0 }
         let tokenValue = CurrencyFormatter.plainFormatter.string(from: token.valueBigInt, decimals: token.decimals).doubleValue
-        return tokenValue * price.priceUSD
+        let priceInUsd = Double(price.price) ?? 0
```
[the remark this answers is not in the document]
Comment (OlegGordiichuk) [reply Trust/Tokens/ViewModels/TokensViewModel.swift]: Done

File: Trust/Tokens/Types/TokensDataStore.swift (hunk 3 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/Types/TokensDataStore.swift]: I think in next iteration and update of the CoinTicker I can add similar approach for parsing.

File: Trust/Settings/Types/Config.swift [hunk 9]
```diff
@@ -17,6 +18,25 @@ struct Config {
     ) {
         self.defaults = defaults
     }
+    
+    var currency: Currency {
```
Comment (michaelScoff) [inline Trust/Settings/Types/Config.swift]: @OlegGordiichuk can you add tests for this? to make sure it works as expected in different cases

**Pull request merged**